Spigot-Server-Patches/0418-ticketlevel-per-player-view-distance-api.patch

```diff
@@ -0,0 +1,184 @@
+ import it.unimi.dsi.fastutil.objects.ObjectSortedSet;
++
++import java.util.Arrays;
+ import java.util.Set;
+ import java.util.concurrent.CompletableFuture;
+ import java.util.concurrent.Executor;
+@@ -352,7 +354,7 @@ public abstract class ChunkMapDistance {
+ 
+         private int e = 0;
+         private final Long2IntMap f = Long2IntMaps.synchronize(new Long2IntOpenHashMap());
+-        private final LongSet g = new LongOpenHashSet();
++        private final Long2ByteMap g = new Long2ByteOpenHashMap(); //Paper ticket level pp-view-dist
+ 
+         protected c(int i) {
+             super(i);
+@@ -360,9 +362,9 @@ public abstract class ChunkMapDistance {
+         }
+ 
+         @Override
+-        protected void a(long i, int j, int k) {
+-            this.g.add(i);
+-        }
++        protected void a(long i, int j, int k, byte d) {
++            this.g.put(i, d);
++        } //Paper ticket level pp-view-dist
+ 
+         public void a(int i) {
+             ObjectIterator objectiterator = this.a.long2ByteEntrySet().iterator();
+@@ -372,20 +374,20 @@ public abstract class ChunkMapDistance {
+                 byte b0 = it_unimi_dsi_fastutil_longs_long2bytemap_entry.getByteValue();
+                 long j = it_unimi_dsi_fastutil_longs_long2bytemap_entry.getLongKey();
+ 
+-                this.a(j, b0, this.c(b0), b0 <= i - 2);
++                this.a(j, b0, this.c(b0), b0 <= i - 2, i);
+             }
+ 
+             this.e = i;
+         }
+ 
+-        private void a(long i, int j, boolean flag, boolean flag1) {
++        private void a(long i, int j, boolean flag, boolean flag1, int vd) { //Paper ticket level pp-view-dist
+             if (flag != flag1) {
+                 Ticket<?> ticket = new Ticket<>(TicketType.PLAYER, ChunkMapDistance.b, new ChunkCoordIntPair(i), ChunkMapDistance.this.currentTick);
+ 
+                 if (flag1) {
+                     ChunkMapDistance.this.j.a(ChunkTaskQueueSorter.a(() -> { // Craftbukkit - decompile error
+                         ChunkMapDistance.this.m.execute(() -> {
+-                            if (this.c(this.c(i))) {
++                            if (this.c(i) <= vd - 2) { //Paper ticket level pp-view-dist
+                                 ChunkMapDistance.this.addTicket(i, ticket);
+                                 ChunkMapDistance.this.l.add(i);
+                             } else {
+@@ -412,10 +414,12 @@ public abstract class ChunkMapDistance {
+         public void a() {
+             super.a();
+             if (!this.g.isEmpty()) {
+-                LongIterator longiterator = this.g.iterator();
++                ObjectIterator<Long2ByteMap.Entry> longiterator = this.g.long2ByteEntrySet().iterator(); //Paper ticket level pp-view-dist
+ 
+                 while (longiterator.hasNext()) {
+-                    long i = longiterator.nextLong();
++                    Long2ByteMap.Entry e = longiterator.next(); //Paper ticket level pp-view-dist
```

Comment:
`// Paper start - Ticket level per player view distance` and `// Paper end`, instead of this